hudi-utilities/src/main/java/org/apache/hudi/utilities/transform/Transformer.java

```diff
@@ -45,4 +47,9 @@ public interface Transformer {
    */
   @PublicAPIMethod(maturity = ApiMaturityLevel.STABLE)
   Dataset<Row> apply(JavaSparkContext jsc, SparkSession sparkSession, Dataset<Row> rowDataset, TypedProperties properties);
+
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  default Option<Schema> transformedSchema(JavaSparkContext jsc, SparkSession sparkSession, Schema incomingSchema, TypedProperties properties) {
+    return Option.empty();
```

Comment:
The default here in my opinion should create an empty dataset with the `incomingSchema` and then apply the transformer and call `.schema()` on the resulting dataset to get the struct type and convert that back to avro. 

Another note, since transforms deal with Rows, does it make more sense to track the schema as a StructType?

Reply:
The default would be the same as tranformed dataset schema then and schema validation would be successful by default. The idea is for API to provide an expected schema and then we verify using it and fail if it is not provided.

Since schema provider provides the schema in Avro format, we are using Avro schema in the API. Just for consistency for the user.

Reply:
I think this puts a large burden on the user to implement these schemas when there is an easier way to get a signal as to whether the combination of transformers is valid.

Reply:
Another note on Avro vs StructType. There are subtle differences between the two and possible struct types that cannot be covered by Avro schemas like non-string map keys. Since this is some intermediate state, I think we should be using the schema format that represents this state so we don't getting into any edge cases where the user cannot define the proper schema of their row before and after the transform

Reply:
Thanks @the-other-tim-brown! These are valid concerns. It would need some more discussion.
cc @rmahindra123 @nsivabalan

Reply:
1. Switch to StructType instead of Avro? @rmahindra123 will confirm
2. Default should infer schema using spark plan